src/thunderbots/software/network_input/filter/robot_filter.cpp

```diff
@@ -2,20 +2,72 @@
 
 #include <ros/console.h>
 
-RobotFilter::RobotFilter(unsigned int id) : robot_id(id) {}
+/**
+ * Implementation of the robot filter
+ *
+ * Helps with resolving robot position when there are mutiple data structs
+ * with the same robot id by averaging, and deals with missing data structs
+ * by caching
+ *
+ */
+
+RobotFilter::RobotFilter(unsigned int id) {}
 
 FilteredRobotData RobotFilter::getFilteredData(
-    const std::vector<SSLRobotData> &new_robot_data)
+    const std::vector<SSLRobotData>& new_robot_data)
 {
-    FilteredRobotData filtered_data;
-    filtered_data.id               = 0;
-    filtered_data.position         = Point();
-    filtered_data.velocity         = Vector();
-    filtered_data.orientation      = Angle::zero();
-    filtered_data.angular_velocity = AngularVelocity::zero();
-    filtered_data.timestamp        = Timestamp::fromSeconds(0);
-
-    return filtered_data;
+    // return cache if there is no data
+    if (new_robot_data.empty())
+    {
+        return cached_robot_data;
+    }
+
+    // new_robot_data is garunteed to contain at east 1 SSLRobotData object
+    FilteredRobotData filtered_robot_data;
+    filtered_robot_data.id        = new_robot_data.front().id;
+    filtered_robot_data.timestamp = new_robot_data.front().timestamp;
+
+    // init position and orientation to zeros
+    filtered_robot_data.position    = Point(0, 0);
+    filtered_robot_data.orientation = Angle::zero();
+
+    // get average position and angle from data
+    for (SSLRobotData data : new_robot_data)
+    {
+        filtered_robot_data.position = filtered_robot_data.position + data.position;
```

Comment:
``` suggestion
        filtered_robot_data.position += data.position;

Reply:
the += operator wasn't working, it adds the value but then doesn't assign it

Reply:
That's odd.... added an issue for this https://github.com/UBC-Thunderbots/Software/issues/340, will try to look at later. Could you leave a quick `TODO: Change to += when #340 is resolved` here?